drivers/misc/optee_rpc.c

```diff
@@ -0,0 +1,366 @@
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/signal.h>
+#include <nuttx/kmalloc.h>
+#include <stdint.h>
+#include "optee.h"
+#include "optee_msg.h"
+#include "optee_rpc.h"
+
+#ifdef CONFIG_DEV_OPTEE_SUPPLICANT
+#  include "optee_supplicant.h"
+#endif
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: optee_rpc_handle_cmd_get_time
+ *
+ * Description:
+ *   Return REE wall-clock time (seconds + nanoseconds) to secure world.
+ *
+ * Parameters:
+ *   arg - Pointer to the RPC message argument located in shared memory by
+ *         the secure world. The answer will be placed in the same argument.
+ *
+ * Returned Value:
+ *   The time is written to:
+ *     arg->params[0].u.value.a   containing seconds since epoch
+ *     arg->params[0].u.value.b   containing nanoseconds
+ *   Result code is written to arg->ret.
+ *
+ ****************************************************************************/
+
+static void optee_rpc_handle_cmd_get_time(FAR struct optee_msg_arg *arg)
+{
+  struct timespec ts;
+
+  /* OP-TEE parameter validation. */
+
+  if (arg->num_params != 1 ||
+      (arg->params[0].attr & OPTEE_MSG_ATTR_TYPE_MASK)
+      != OPTEE_MSG_ATTR_TYPE_VALUE_OUTPUT)
+    {
+      arg->ret = TEE_ERROR_BAD_PARAMETERS;
+      return;
+    }
+
+  if (clock_gettime(CLOCK_REALTIME, &ts) < 0)
+    {
+      /* Should not happen unless the RTC driver is missing */
+
+      arg->ret = TEE_ERROR_GENERIC;
+      return;
+    }
+
+  arg->params[0].u.value.a = (uint32_t)ts.tv_sec;   /* Seconds since epoch. */
+  arg->params[0].u.value.b = (uint32_t)ts.tv_nsec;  /* Nanoseconds.         */
+
+  arg->ret = TEE_SUCCESS;
+  return;
+}
+
+/****************************************************************************
+ * Name: optee_rpc_cmd_suspend
+ *
+ * Description:
+ *   Request from OP-TEE to suspend the current nuttx process.
+ *
+ * Parameters:
+ *   arg - Pointer to the RPC message argument, located in a shared page, by
+ *         the secure world, containing the time in msec to sleep.
+ *
+ * Returned Value:
+ *   None.  Result codes are written into arg->ret.
+ *
+ ****************************************************************************/
+
+static void optee_rpc_cmd_suspend(FAR struct optee_msg_arg *arg)
+{
+  /* OP-TEE parameter validation. */
+
+  if (arg->num_params != 1 ||
+      (arg->params[0].attr & OPTEE_MSG_ATTR_TYPE_MASK) !=
+          OPTEE_MSG_ATTR_TYPE_VALUE_INPUT)
+    {
+      arg->ret = TEE_ERROR_BAD_PARAMETERS;
+      return;
+    }
+
+  uint32_t msec_to_wait = arg->params[0].u.value.a;
```

Comment:
uint32_t to useconds_t, move the declaration to the begin of fucntion and remove the cast at line 120